### ISSUE TITLE
Potential issues with command checker interrupter

### ISSUE THREAD

**Issue by doublep** [flycheck/flycheck#1791]
As far as I understand it, command checker interrupter just kills processes outright. Moreover, there is no "official" way to disable or replace it (I know how to do it hackishly, but that's not the point).

This might be not be the best idea for some processes. Here I'm speaking about Eldev, of course, which is basically a Elisp program that uses Emacs package manager to install project dependencies on-the-fly. This is often a lengthy process, taking anywhere from 3 seconds to a couple of minutes if there are *many* dependencies. If Emacs is killed in the middle of package installation, I'm sure `package-user-dir` is left in an inconsistent state. 

There might be other commands that are not too happy about being killed violently.

I admit I don't have too many ideas what exactly to do, I guess at least half of improvements have to be done on Eldev side. But I'm sure Flycheck has some space for improvement here as well.

For starters, I would propose two things:

* If I understand the source code of Flycheck and Emacs correctly, processes are really *killed* (SIGKILL). I think SIGTERM or SIGINT would be more suitable.
* Open up official possibility to pass `:interrupt` to `flycheck-define-command-checker`.

**Comment by cpitclaudel** [comment]
Indeed, passing SIGTERM would be more reasonable (with a timer set to send a SIGKILL after a few seconds, though).

> Moreover, there is no "official" way to disable or replace it (I know how to do it hackishly, but that's not the point).

I think it's quite fine to replace the `:interrupt` on a command checker.  Concretely, what would you use it for on the eldev side, currently?

(Btw, I think installing new packages when checking with eldev should likely be an option (eldev could return an error otherwise, saying packages need to be installed).  It's a bit scary that just opening a file can start downloading stuff)

**Comment by doublep** [comment]
> I think it's quite fine to replace the :interrupt on a command checker. Concretely, what would you use it for on the eldev side, currently?

Currently I don't know. I would just like to have an option to use something else (ideally with a simple way to fall back to the original behavior) in case I do want it at some point.

> (Btw, I think installing new packages when checking with eldev should likely be an option (eldev could return an error otherwise, saying packages need to be installed). It's a bit scary that just opening a file can start downloading stuff)

I guess you have a point, especially given that it can download from anywhere (just like Cask, for that matter). It's one thing when you do it explicitly — e.g. even `make` could screw you if you run it, — and another when it is triggered just by opening a file. I will think about this.

**Comment by doublep** [comment]
Thinking further, it doesn't matter much. Emacs Lisp checker already byte-compiles files and byte-compiling can invoke arbitrary code already, e.g. by using `(eval-when-compile ...)`. It could also download something or erase your `~` etc. With Eldev it is even marginally safer, because Eldev isolates the projects from your normal Emacs (though, of course, given that it also runs arbitrary code by the virtue of reusing Emacs Lisp checker, this can be easily circumvented).

What would _really_ make things safer is running Emacs Lisp checker (and also half of other checkers, I guess) in a fully sandboxed environment, but that doesn't sound simple to achieve.

**Comment by cpitclaudel** [comment]
> Currently I don't know. I would just like to have an option to use something else (ideally with a simple way to fall back to the original behavior) in case I do want it at some point.

I'm a bit wary of extending the API without having a concrete use case is mind, but I do agree that sending a term before a kill is the right thing to do.

> What would really make things safer is running Emacs Lisp checker (and also half of other checkers, I guess) in a fully sandboxed environment, but that doesn't sound simple to achieve.

Yes, that would be quite tricky.  Checker security is a long-standing todo.  Still, when the issue came up with the Haskell stack checker, we turned off the auto-installation of dependencies.  I tihk having a setting could be good.

**Comment by doublep** [comment]
> I'm a bit wary of extending the API without having a concrete use case is mind

Well, it wouldn't really extend it, you can already pass `:interrupt` when defining a non-command based checker. Also, `flycheck-interrupt-command-checker` is already non-package-private, so there is an option to use the standard fallback easily. I think just removing the guard against specifying `:interrupt` in `flycheck-define-command-checker` and its mention in documentation (or rewording appropriately) would be enough and also not a huge change.

> Still, when the issue came up with the Haskell stack checker, we turned off the auto-installation of dependencies.

I know practically nothing about Haskell, but I suspect for Haskell this was enough (if it doesn't evaluate during compiling), unlike for Emacs Lisp checker. I will think about this, though.

**Comment by bbatsov** [comment]
Closing: feature request specific to `flycheck-eldev` extension's needs. Stale for 5+ years.
